Add information about how existing in-trees PVs and in-tree format yaml file will be handled after CSI migration is completed (#44295)
kubernetes/website: content/en/docs/concepts/storage/volumes.md

for a related mechanism). Also, a volume cannot contain a hard link to anything in
a different volume.

> ## In-tree volume plugins and their CSI Migration progress {#volume-types}

Reviewer: Since it is explained below, I don't think you need to change the title.

> Below is a list of In-tree plugins that are supported by Kubernetes and the status of their respective CSI Migrations progress.

Author: ```suggestion
Below is a list of in-tree plugins that are supported by Kubernetes and the status of their respective CSI Migrations progress.
```

> The Container Storage Interface (CSI) was designed to help Kubernetes replace its existing, in-tree storage driver mechanisms - especially vendor specific plugins.Kubernetes support for the Container Storage Interface has been generally available since Kubernetes v1.13. For more details on CSI Migration, please refer to [Out-of-tree volume plugins](#out-of-tree-volume-plugins).

Author: ```suggestion
The Container Storage Interface (CSI) was designed to help Kubernetes replace its existing, in-tree storage driver mechanisms - especially vendor specific plugins. Kubernetes support for the Container Storage Interface has been generally available since Kubernetes v1.13. For more details on CSI Migration, please refer to [Out-of-tree volume plugins](#out-of-tree-volume-plugins).
```

> The Container Storage Interface (CSI) was designed to help Kubernetes replace its existing, in-tree storage driver mechanisms - especially vendor specific plugins. Kubernetes support for the Container Storage Interface has been generally available since Kubernetes v1.13. For more details on CSI Migration, please refer to [Out-of-tree volume plugins](#out-of-tree-volume-plugins).

Reviewer: This content is written  in `[Out-of-tree volume plugins](#out-of-tree-volume-plugins)`, so I think it is enought to say 'For more details...'.

{{< note >}}

> When a Kubernetes cluster administrator updates a cluster to enable CSI migration, existing workloads that utilize PVCs which are backed by in-tree storage plugins will continue to function as they always have. Existing In trees PVs and In Tree format yaml file will still be used in the future even after the migration is completed, and will not be left unprocessed or deleted in the future. Users can still specify the following In-tree plugins in their yaml file. However, behind the scenes, Kubernetes hands control of all storage management operations (previously targeting in-tree drivers) to CSI drivers.

Author: ```suggestion
When a Kubernetes cluster administrator updates a cluster to enable CSI migration, existing workloads that utilize PVCs which are backed by in-tree storage plugins will continue to function as they always have. Existing in-tree PVs and in-tree format yaml file will still be used in the future even after the migration is completed, and will not be left unprocessed or deleted in the future. Users can still specify the following in-tree plugins in their yaml file. However, behind the scenes, Kubernetes hands control of all storage management operations (previously targeting in-tree drivers) to CSI drivers.
```

> When a Kubernetes cluster administrator updates a cluster to enable CSI migration, existing workloads that utilize PVCs which are backed by in-tree storage plugins will continue to function as they always have. Existing in-tree PVs and in-tree format yaml file will still be used in the future even after the migration is completed, and will not be left unprocessed or deleted in the future. Users can still specify the following in-tree plugins in their yaml file. However, behind the scenes, Kubernetes hands control of all storage management operations (previously targeting in-tree drivers) to CSI drivers.

Reviewer: Would you consider expressing it more concisely as follows?

If you switch that plugin to an out-of-tree one in the Kubernetes configuration, or if the plugin is removed from the in-tree, the workload can still use the existing manifest yaml file or the PV created without any changes. This behavior is ensured by the migrated out-of-tree plug-in.

configuration changes to existing Storage Classes, PersistentVolumes or PersistentVolumeClaims
(referring to in-tree plugins) when transitioning to a CSI driver that supersedes an in-tree plugin.

{{< note >}}

> * The existing PV created by in-tree volume plugin can still be used in the future without any configuration changes even after the `CSIMigration` is completed. 

Reviewer 2: - "The" looks unnecessary: `The existing PV` -> `Existing PVs`
- I think it is better to add `and a corresponding in-tree volume plugin is removed` after `after the CSIMigration is completed` for further clarification. While I know the scope of CSI migration effort includes removals of corresponding in-tree volume plugins, I still think this kind of clarification helps operators who run kubernetes.

```suggestion
* Existing PVs created by in-tree volume plugin can still be used in the future without any configuration changes even after the `CSIMigration` is completed and corresponing in-tree volume plugins are deleted.
```

Reviewer 3: Good point, I'm also agree this suggestion.

> * A new PV with a manifest that uses in-tree volume plugins can still be created, even after the in-tree volume plugin is removed from Kubernetes.

Reviewer 2: nit: After CSI migration effort completes and a corresponding in-tree volum plugin is removed, operations on PVs are actually done by CSI drivers, so how about `refers to` instead of `(a manifest that) uses (in-tree volume plugins ...)`.

another nit: a single PV cannot refer to multiple in-tree volume plugins, so it is better to use singular/plural cases consistently, like `new PVs refering to in-tree volume plugins`.

Author: Thank you for you kind suggestion! I've corrected the second sentence as you pointed out.

> * New PVs with manifests that referring to in-tree volume plugins can still be created, even after the in-tree volume plugins are removed from Kubernetes.

Reviewer 2: nit: `that referrig to` -> `referring to` or `that refer to`
Perhaps my last comment suggested it :-(

another nit:: a manifest includes PV(s) so, I personally prefer to "PVs in manifests" over "with manifests".

Author: Thank you for you kind suggestion! I've corrected the second sentence as you pointed out.

configuration changes to existing Storage Classes, PersistentVolumes or PersistentVolumeClaims
(referring to in-tree plugins) when transitioning to a CSI driver that supersedes an in-tree plugin.

{{< note >}}

> * Existing PVs created by in-tree volume plugin can still be used in the future without any configuration changes even after the `CSIMigration` is completed and corresponing in-tree volume plugins are deleted.

Reviewer 4: Meaning is fine, but working as a typo and may mislead.

Try this:
```suggestion
Existing PVs created by a in-tree volume plugin can still be used in the future without any configuration
changes, even after the migration to CSI is completed for that volume type, and even after you upgrade to a
version of Kubernetes that doesn't have compiled-in support for that kind of storage.
 
As part of that migration, you - or another cluster administrator - **must** have installed and configured
the appropriate CSI driver for that storage. The core of Kubernetes does not install that software for you.
```

> * New PVs in manifests referring to in-tree volume plugins can still be created, even after the in-tree volume plugins are removed from Kubernetes.

Reviewer 4: Try this:
```suggestion

---

After that migration, you can also define new PVCs and PVs that refer to the legacy, built-in
storage integrations.
Provided you have the appropriate CSI driver installed and configured, the PV creation continues
to work, even for brand new volumes. The actual storage management now happens through
the CSI driver.
```
